tests/python_package_test/test_plotting.py

```diff
@@ -188,6 +203,37 @@ def test_create_tree_digraph(breast_cancer_split):
     assert 'count' not in graph_body
 
 
+@pytest.mark.skipif(not GRAPHVIZ_INSTALLED, reason='graphviz is not installed')
+def test_tree_with_categories_below_max_category_values(categorical_data):
+    X_train, _, y_train, _ = categorical_data(2, 10)
+    gbm = lgb.LGBMClassifier(n_estimators=10, num_leaves=3, verbose=-1)
```

Comment:
> but the problem is the way that lightgbm splits on categorical features is not deterministic

This is not true. It's possible to get `lightgbm` to deterministically produce the same model given the same input data... it's just not deterministic *by default*.

You can use the following code (pay special attention to the values of `params`) as an example.

```python
import lightgbm as lgb
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer
X, y = load_breast_cancer(return_X_y=True)

X_df = pd.DataFrame()
rnd = np.random.RandomState(0)
n_cat_values = rnd.randint(2, 10, size=X.shape[1])
for i in range(X.shape[1]):
    bins = np.linspace(0, 1, num=n_cat_values[i]+1)
    X_df[f"cat_col_{i}"] = pd.qcut(X[:, i], q=bins, labels=range(n_cat_values[i])).as_unordered()

params = {
    "n_estimators": 10,
    "num_leaves": 3,
    "min_data_in_bin": 1,
    "force_col_wise": True,
    "deterministic": True,
    "num_threads": 1,
    "seed": 708,
    "verbose": -1
}

gbm1 = lgb.LGBMClassifier(**params)
gbm1.fit(X_df, y)
gbm1_df = gbm1.booster_.trees_to_dataframe()

# prove that in 100 repeated training runs, the exact same model is produced
for i in range(100):
    print(f"iteration: {i}")
    gbm = lgb.LGBMClassifier(**params).fit(X_df, y)
    pd.testing.assert_frame_equal(gbm1_df, gbm.booster_.trees_to_dataframe())
```

Try modifying your tests to pass similar values of `params`, to produce the same model every time the tests are run.

Reply:
That is interesting, I did not know that ```deterministic``` parameter exists